src/pykeen/nn/pyg.py

```diff
@@ -0,0 +1,290 @@
+        """
+        Initialize the representation.
+
+        :param triples_factory:
+            the factory comprising the training triples used for message passing
+        :param kwargs:
+            additional keyword-based parameters passed to :meth:`AbstractPyGRepresentation.__init__`
+        """
+        super().__init__(triples_factory=triples_factory, **kwargs)
+        self.register_buffer(name="edge_type", tensor=triples_factory.mapped_triples[:, 1])
+
+    def _message_passing(self, x: torch.FloatTensor) -> torch.FloatTensor:  # noqa: D102
+        for layer, activation in zip(self.layers, self.activations):
+            x = activation(layer(x, edge_index=self.edge_index, edge_type=self.edge_type))
+        return x
+
+
+class FeaturizedRelationTypePyGRepresentation(CategoricalRelationTypePyGRepresentation):
+    """A representation with message passing with uses categorical relation type information, e.g., R-GCN."""
+
+    #: the edge type
+    relation_representation: Representation
+
+    def __init__(
+        self,
+        triples_factory: CoreTriplesFactory,
+        relation_representation: HintOrType[Representation] = None,
+        relation_representation_kwargs: OptionalKwargs = None,
+        relation_transformation: Optional[nn.Module] = None,
+        **kwargs,
+    ):
+        """
+        Initialize the representation.
+
+        :param triples_factory:
+            the factory comprising the training triples used for message passing
+
+        :param relation_representation:
+            the base representations for relations, or a hint thereof
+        :param relation_representation_kwargs:
+            additional keyword-based parameters passed to the base representations upon instantiation
+
+        :param relation_transformation:
+            an optional transformation to apply to the relation representations after each message passing step.
+            If None, do not modify the representations.
+
+        :param kwargs:
+            additional keyword-based parameters passed to :meth:`AbstractPyGRepresentation.__init__`
+        """
+        super().__init__(triples_factory=triples_factory, **kwargs)
+        # avoid cyclic import
+        from . import representation_resolver
+
+        self.relation_representation = representation_resolver.make(
+            relation_representation, pos_kwargs=relation_representation_kwargs, max_id=triples_factory.num_relations
+        )
+        self.relation_transformation = relation_transformation
+
+    def _message_passing(self, x: torch.FloatTensor) -> torch.FloatTensor:  # noqa: D102
+        edge_attr = self.relation_representation(self.edge_type)
+        for layer, activation in zip(self.layers, self.activations):
+            x = activation(layer(x, edge_index=self.edge_index, edge_attr=edge_attr))
+            if self.relation_transformation is not None:
+                edge_attr = self.relation_transformation(edge_attr)
```

Comment:
slight performance improvement from not running this check on the last layer

Reply:
5cfb253